### PR TITLE
Change sharing map so that leafs can be directly attached to internal nodes

### PR DESCRIPTION
**Pull request by danpoe**
Previously all leafs were stored in a linked list. However, this was unnecessary as collisions can occur only between nodes at the bottom of the tree. Now only leafs at the bottom are stored in a linked list, and all other leafs are direct successors of internal nodes. 

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by danpoe** [comment]
On a set of methods from Apache Tika, the sum of the runtimes of the sharing map methods (timing just those) improved by 28%. Overall symex was 7% faster.

File: src/util/sharing_map.h [hunk 1]
```diff
@@ -193,10 +193,8 @@ class sharing_mapt
   typedef std::vector<key_type> keyst;
 
 protected:
-  typedef sharing_node_innert<key_type, mapped_type> innert;
-  typedef sharing_node_leaft<key_type, mapped_type> leaft;
-
-  typedef sharing_node_baset baset;
+  typedef sharing_nodet<key_type, mapped_type> innert;
```
Comment (smowton) [inline src/util/sharing_map.h]: (random line) commit message: where -> were :p

File: src/util/sharing_node.h [hunk 2]
```diff
@@ -70,14 +69,14 @@ const T *as_const(T *t)
 
 // Inner nodes (internal nodes or container nodes)
```
Comment (smowton) [inline src/util/sharing_node.h]: Comment out of date now there are 3 possibilities?

File: src/util/sharing_map.h [hunk 3]
```diff
@@ -518,20 +518,8 @@ class sharing_mapt
...
-    const innert *cp = get_container_node(k);
-    if(cp == nullptr)
-      return nullptr;
-
-    const leaft *lp;
-    lp = cp->find_leaf(k);
-
-    return lp;
-  }
+  leaft *get_leaf_node(const key_type &k);
```
Comment (smowton) [inline src/util/sharing_map.h]: (random line) commit message: leafs -> leaf

File: src/util/sharing_map.h [hunk 4]
```diff
@@ -1077,10 +1079,21 @@ SHARING_MAPT2(const, innert *)::get_container_node(const key_type &k) const
...
-    if(ip->is_container())
-      return ip;
+    if(ip->is_internal())
+    {
+      key >>= chunk;
+      continue;
+    }
 
-    key >>= chunk;
+    if(ip->is_leaf())
```
Comment (smowton) [inline src/util/sharing_map.h]: These are mutually exclusive, right? If so use `else if` to clarify that.

File: src/util/sharing_map.h [hunk 5]
```diff
@@ -1111,62 +1124,72 @@ SHARING_MAPT(void)::erase(const key_type &k)
...
-    SM_ASSERT(!ip->empty());
+    PRECONDITION(!ip->empty());
+
+    if(ip->is_internal())
+    {
+      key >>= chunk;
+      continue;
+    }
+
+    if(ip->is_leaf())
```
Comment (smowton) [inline src/util/sharing_map.h]: `else if`

File: src/util/sharing_map.h [hunk 6]
```diff
@@ -1111,62 +1124,72 @@ SHARING_MAPT(void)::erase(const key_type &k)
...
+    {
+      PRECONDITION(equalT()(ip->get_key(), k));
+      del->remove_child(del_bit);
+
+      num--;
+
+      return;
+    }
 
     if(ip->is_container())
```
Comment (smowton) [inline src/util/sharing_map.h]: `else if`

**Review by smowton — APPROVED** [review]
LGTM (at moderate depth, haven't walked through the algorithms in detail to make sure there are no missing cases). Surprised not to see any updates to the top-level docs for the data structure though?

File: src/util/sharing_map.h [hunk 7]
```diff
@@ -1233,88 +1299,93 @@ ::insert(const key_type &k, valueU &&m)
...
-  innert *cp = get_container_node(k);
-  SM_ASSERT(cp != nullptr);
-
-  leaft *lp = cp->find_leaf(k);
-  PRECONDITION(lp != nullptr); // key must exist in map
+  nodet &lp = get_leaf_node(k);
 
   INVARIANT(
-    !value_equalt()(as_const(lp)->get_value(), m),
+    !value_equalt()(as_const(lp).get_value(), m),
```
Comment (tautschnig) [inline src/util/sharing_map.h]: Visual Studio complains that the use of `as_const` is ambiguous here. You might have add/use a static cast here.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e7a5cf1d7dd53ebc51072e994b92f60d5ca96cf1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112914521
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 872f2030df952458fd0472fc25e4c6ba685dc1b1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112923269

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d43834165ffef151630e3a5246039ef1463a9a54).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112940281

**Pull request merged**